### DISCUSSION TITLE
[Feature] Publish packages to NPM via OIDC

### DISCUSSION THREAD

**Discussion by mxschmitt** [community/community#127011] [npm] [answered]
Motivation: Authentication flows are more and more secret-free / secret-less. [OIDC](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect) provides a way to say in simple terms: `allow publishing from GitHub repository github.com/foo/bar with branch 'main' to a new NPM version`.

This is similar to how the authentication flow from [GitHub Actions to Azure](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-azure) works and is recommended nowadays. There you set the OIDC subject to a similar value.

Similar to https://github.com/npm/feedback/discussions/872

Scope: npmjs.com

**Comment by Shegox** [comment]
I fully support this suggestion. Enabling OIDC federation between GitHub.com and npmjs.com would be incredibly beneficial, similar to how PyPI handles it: [Configuring OpenID Connect in PyPI](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-pypi).

Our specific use case involves multiple repositories managed by different teams, all releasing packages under the same scope on npmjs. To manage permissions, we currently create granular access tokens and distribute them to the GitHub Actions Secrets of individual repositories. Since these granular access tokens cannot be created via API, the process must be done manually, including the manual rotation of these tokens, which is cumbersome.

Additionally, I would like to highlight a notable feature of the PyPI setup: the ability to configure OIDC for a not yet existing package. With npmjs.com, you must first create a dummy version before generating a granular access token for the package. If OIDC is implemented, it would be fantastic if we could directly set it up for packages that don't yet exist.

**Reply by harshit-kochar** [reply]
@leobalter Any plans on releasing the ability to configure OIDC for not-yet-existing packages?

**Reply by 34j** [reply]
@harshit-kochar as a workaround, [setup-npm-trusted-publish](https://github.com/azu/setup-npm-trusted-publish), a dedicated tool for this, can be used to publish a *"dummy"* package to npm:

```shell
npx --yes setup-npm-trusted-publish <package-name>
```

**Comment by janbrasna** [comment]
I would even go as far as suggesting that npmjs.com should be able to understand the actual GITHUB_TOKEN if granted `packages:write` and allow setting the publishing or connecting the package and its repo in the same way as the GH registry does.

I would have expected more integration since npm has been acquired some five years ago. (In my use-case, I'd still prefer to use npmjs registry for continuity instead of pkg.gh, but am not really willing to manage isolated users, permissions, accounts etc. separately there — and would love to just link the npmjs packages to the gh repos, and let the access controls there do their magic regarding teams, granular permissions, automation tokens etc. — eventually be just able to publish with GITHUB_TOKEN in CI…)

**Comment by 0xGorilla** [comment]
How can we help move this forward? The current publishing workflow is quite dangerous, I think we can all agree on that one.

**Reply by leobalter** [reply]
I don't wanna break the spoilers here but we are actively working on this feature. I should have a post with timelines coming up in the next couple weeks.

**Reply by turtlemoji** [reply]
this is great news ❤️

**Reply by aripalo** [reply]
This is fantastic news! 🤩 I have a lot of workflows that dual publish to both GitHub Packages and npmjs.com. Having to manage the npmjs.com access tokens is kinda PITA (especially dealing with rotation) and potential security concern if done wrong.

**Reply by lastmjs** [reply]
@leobalter Do you have a link to that post with timelines?

**Reply by lastmjs** [reply]
I assume this is it? https://github.com/orgs/community/discussions/161015

**Comment by Shegox** [comment]
Just FYI for anyone watching and missing the announcement: 
- Announcement: https://github.com/orgs/community/discussions/161015
- PR for the npm CLI: https://github.com/npm/cli/pull/8336

The relevant parts: 
> At launch, we'll support package publishing from:
> -  GitHub Actions workflows
> -  GitLab CI/CD pipelines
> 
> 🎯 Private Preview: June 2025 - Limited access for early testers
> 🌍 Public Release: July 2025 (tentative) - General availability for all users

**Reply by lastmjs** [reply]
Hey! Any update on the release of general availability?

**Reply by leobalter** [reply, marked as answer]
Hi @lastmjs I'm sorry I didn't answer yesterday but [this new update](https://github.blog/changelog/2025-07-31-npm-trusted-publishing-with-oidc-is-generally-available/) might be helpful.

Trusted publishing with OIDC is now generally available!

**Reply by lastmjs** [reply]
This timing is amazing! I literally just became aware of needing the feature yesterday! Good work!

**Reply by twesterhuys** [reply]
In the light of the recent npm supply chain attacks: Is there any plan to support this from self-hosted runners/GHES?

**Comment by tats-u** [comment]
How can we publish the first version with a provenance without NPM_TOKEN or `@namespace/` from GitHub Actions?

**Reply by turtlemoji** [reply]
I think you can't, what I do is publish a placeholder with a simple package.json first manually.

You can't also setup multiple deploy actions or environments.

**Reply by turtlemoji** [reply]
```
{
  "name": "@<your-org>/<package-name>",
  "version": "0.0.0",
  "description": "Placeholder package. Real code coming soon.",
  "license": "MIT",
  "main": "index.js",
  "types": "index.d.ts",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "publish:park": "npm publish --tag snapshot --access public"
  }
}
```

**Reply by tats-u** [reply]
Ended up with a short-lived npm token allowing write to all packages.
